src/reducers/columnStateHelper.js

```diff
@@ -47,79 +42,180 @@ function initialize(state, props, oldProps) {
   columnResizingData = isColumnResizing ? columnResizingData : {};
 
   return Object.assign({}, state, {
+    columnAnchor,
     columnResizingData,
     isColumnResizing,
     maxScrollX,
     scrollX,
   });
-};
+}
 
 /**
+ * Get the anchor for scrolling.
+ * This will either be the first columns's index and an offset, or the last columns's index.
+ * We also pass a flag indicating if the anchor has changed from the state
+ *
  * @param {!Object} state
- * @param {{
- *   scrollToColumn: number,
- *   width: number,
- * }} props
- * @param {number} oldScrollToColumn
- * @param {number} scrollX
- * @return {number} The new scrollX
+ * @param {!Object} props
+ * @param {!Object} oldProps
+ * @return {{
+ *   firstIndex: number,
+ *   firstOffset: number,
+ *   lastIndex: number,
+ *   changed: boolean,
+ * }}
  */
-function scrollTo(state, props, oldScrollToColumn, scrollX) {
-  const { scrollToColumn } = props;
-  if (isNil(scrollToColumn)) {
-    return scrollX;
+function getColumnAnchor(state, props, oldProps) {
+  // if scrollToColumn changed
+  if (props.scrollToColumn !== undefined &&
+    props.scrollToColumn !== null &&
+    (!oldProps || props.scrollToColumn !== oldProps.scrollToColumn)
+  ) {
+    return scrollToColumn(state, props.scrollToColumn);
   }
 
+  // if scrollLeft changed
+  if (
+    props.scrollLeft !== undefined &&
+    props.scrollLeft !== null &&
+    (!oldProps || props.scrollLeft !== oldProps.scrollLeft)
+  ) {
+    return scrollToPos(state, props.scrollLeft);
+  }
+
+  // no change in column anchor
+  return {
+    firstIndex: state.firstColumnIndex,
+    firstOffset: state.firstColumnOffset,
+    lastIndex: undefined,
+    changed: false,
+  };
+}
+
+/**
+ * @param {!Object} state
+ * @param {number} scrollToColumn
+ * @return {{
+ *   firstIndex: number,
+ *   firstOffset: number,
+ *   lastIndex: number,
+ *   changed: boolean,
+ * }}
+ */
+function scrollToColumn(state, scrollToColumn) {
   const {
     availableScrollWidth,
     fixedColumns,
     scrollableColumns,
   } = columnWidths(state);
-  const fixedColumnsCount = fixedColumns.length;
-  const scrollableColumnsCount = scrollableColumns.length;
-
-  const noScrollableColumns = scrollableColumnsCount === 0;
-  const scrollToUnchanged = scrollToColumn === oldScrollToColumn;
-  const selectedColumnFixed = scrollToColumn < fixedColumnsCount;
-  const selectedColumnFixedRight = scrollToColumn >= fixedColumnsCount + scrollableColumnsCount;
-  if (scrollToUnchanged || selectedColumnFixed || selectedColumnFixedRight || noScrollableColumns) {
-    return scrollX;
+
+  const { scrollX, columnOffsetIntervalTree } = state;
+
+  const selectedColumnFixed = scrollToColumn < fixedColumns.length;
+  const selectedColumnFixedRight = scrollToColumn >= fixedColumns.length + scrollableColumns.length;
```

Comment:
Why not retain the variables for the column count which is more readable?

Reply:
> Why not retain the variables for the column count which is more readable?

sorry, didn't get you

Reply:
I think @somavara is asking why you removed `fixedColumnsCount` and `scrollableColumnsCount`, only to use `fixedColumn.length` and `scrollableColumns.length` here and below.